Add a new note as a placeholder for UPE CTA

harriswong opened the pull request:
Fixes https://github.com/Automattic/woocommerce-payments/issues/2168

#### Changes proposed in this Pull Request
This PR adds an admin note when `_wcpay_feature_upe_settings_preview` is enabled but `_wcpay_feature_upe` is disabled. 

When clicked "Enable on your store", it should enable the UPE feature flag by setting `_wcpay_feature_upe` to 1.

#### Testing instructions
1. Enable UPE settings preview by setting `_wcpay_feature_upe_settings_preview` to 1 or make [this function](https://github.com/Automattic/woocommerce-payments/blob/06ad35edec56d92e87d6d2f9a22bf39b81ffc403/includes/class-wc-payments-features.php#L67) return true.
2. Make sure `_wcpay_feature_upe` doesn't exist in your DB. You can check with `SELECT * FROM `wp_options` where option_name = '_wcpay_feature_upe'`
3. Go to Home `wp-admin/admin.php?page=wc-admin` or Orders `wp-admin/edit.php?post_type=shop_order`, click on "Inbox" and you should see the admin note:
![image](https://user-images.githubusercontent.com/572862/123132221-ba48e400-d40b-11eb-9b74-75b32a4d6ff4.png)

4. Click "Enable on your store", check database using query in step 2 and to confirm the feature flag is added. It should redirects you to `wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments`. 
5. If you would like, you can repeat the above steps for the matrix below:

| _wcpay_feature_upe_settings_preview | _wcpay_feature_upe | Show Admin Note?|
|--|--|--|
|true|true| no|
|true|false| yes |
|false|true| no|
|false|false| no|

### TODO:
- Update learn more link, pending. We can make a follow-up issue

-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

<!--
Please read P7bbVw-3ww-p2 before opening the PR, assign the correct status to it, __and make sure that the related issue uses the Has PR status!__.
-->

frosso commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -364,6 +364,8 @@ public function __construct(
 		// Update the current request logged_in cookie after a guest user is created to avoid nonce inconsistencies.
 		add_action( 'set_logged_in_cookie', [ $this, 'set_cookie_on_current_request' ] );
+
+		add_action( 'admin_init', [ $this, 'maybe_enable_upe_feature_flag' ] );
```
It might be nice to colocate this in the note itself, like we discussed in the team meeting. In the main comment, I added the diff that has been working for me.

frosso reviewed (commented):
I do think it would be nice to colocate this within the note, like we chatted at the team meeting.

For simplicity, here's the changes the would make it work (including some of the suggestions below).

- I added an `init` method in the note, which gets called instead of `possibly_add_note`
- The `init` method also checks for the action & does the redirect

```diff
diff --git a/includes/class-wc-payment-gateway-wcpay.php b/includes/class-wc-payment-gateway-wcpay.php
index e54508af..5312d14a 100644
--- a/includes/class-wc-payment-gateway-wcpay.php
+++ b/includes/class-wc-payment-gateway-wcpay.php
@@ -364,8 +364,6 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 
 		// Update the current request logged_in cookie after a guest user is created to avoid nonce inconsistencies.
 		add_action( 'set_logged_in_cookie', [ $this, 'set_cookie_on_current_request' ] );
-
-		add_action( 'admin_init', [ $this, 'maybe_enable_upe_feature_flag' ] );
 	}
 
 	/**
@@ -2229,30 +2227,4 @@ class WC_Payment_Gateway_WCPay extends WC_Payment_Gateway_CC {
 			]
 		);
 	}
-
-	/**
-	 * Enable UPE feature flag. CTA from Admin Notes. See WC_Payments_Notes_Additional_Payment_Methods.
-	 */
-	public function maybe_enable_upe_feature_flag() {
-		$nonce = '';
-		if ( ! empty( $_GET['_wpnonce'] ) ) {
-			$nonce = sanitize_key( $_GET['_wpnonce'] );
-		}
-
-		if (
-			! wp_verify_nonce( $nonce, 'enable-upe' ) ||
-			empty( $_GET['page'] ) ||
-			'wc-settings' !== $_GET['page'] ||
-			empty( $_GET['action'] ) ||
-			'enable-upe' !== $_GET['action']
-		) {
-			return;
-		}
-
-		// Enable UPE.
-		WC_Payments_Features::update_upe_enabled( true );
-
-		$wcpay_settings_url = admin_url( 'admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments' );
-		wp_safe_redirect( $wcpay_settings_url );
-	}
 }
diff --git a/includes/class-wc-payments.php b/includes/class-wc-payments.php
index 5ae0ef33..60f37aa3 100644
--- a/includes/class-wc-payments.php
+++ b/includes/class-wc-payments.php
@@ -793,7 +793,7 @@ class WC_Payments {
 			WC_Payments_Notes_Set_Https_For_Checkout::possibly_add_note();
 
 			require_once WCPAY_ABSPATH . 'includes/notes/class-wc-payments-notes-additional-payment-methods.php';
-			WC_Payments_Notes_Additional_Payment_Methods::possibly_add_note();
+			WC_Payments_Notes_Additional_Payment_Methods::init();
 		}
 	}
 
diff --git a/includes/notes/class-wc-payments-notes-additional-payment-methods.php b/includes/notes/class-wc-payments-notes-additional-payment-methods.php
index 28e3a9f6..4afd4318 100644
--- a/includes/notes/class-wc-payments-notes-additional-payment-methods.php
+++ b/includes/notes/class-wc-payments-notes-additional-payment-methods.php
@@ -21,18 +21,64 @@ class WC_Payments_Notes_Additional_Payment_Methods {
 	 */
 	const NOTE_NAME = 'wc-payments-notes-additional-payment-methods';
 
+	/**
+	 * Name of the nonce added to the "enable UPE" action.
+	 */
+	const NONCE_NAME = 'enable-upe';
+
+	/**
+	 * Name of the action used in the URL to enable UPE.
+	 */
+	const ACTION_NAME = 'wcpay-enable-upe';
+
+	/**
+	 * Initializes adding note and possible redirect to enable the feature flag.
+	 */
+	public static function init() {
+		// Show this notice only if UPE settings preview is disabled, and UPE flag is not enabled.
+		if ( ! WC_Payments_Features::is_upe_settings_preview_enabled() ) {
+			return;
+		}
+
+		if ( WC_Payments_Features::is_upe_enabled() ) {
+			return;
+		}
+
+		self::maybe_enable_upe_feature_flag();
+		self::possibly_add_note();
+	}
+
+
+	/**
+	 * Enable UPE feature flag. CTA from Admin Notes. See WC_Payments_Notes_Additional_Payment_Methods.
+	 */
+	public static function maybe_enable_upe_feature_flag() {
+		if (
+			empty( $_GET['page'] ) ||
+			'wc-settings' !== $_GET['page'] ||
+			empty( $_GET['action'] ) ||
+			self::ACTION_NAME !== $_GET['action']
+		) {
+			return;
+		}
+
+		check_admin_referer( self::NONCE_NAME );
+
+		// Enable UPE.
+		update_option( '_wcpay_feature_upe', '1' );
+		self::possibly_delete_note();
+
+		wp_safe_redirect( admin_url( 'admin.php?page=wc-admin&task=woocommerce-payments--additional-payment-methods' ) );
+		exit;
+	}
+
 	/**
 	 * Get the note.
 	 */
 	public static function get_note() {
-		// Show this notice only if UPE settings preview is disabled, and UPE flag is not enabled.
-		if ( ! WC_Payments_Features::is_upe_settings_preview_enabled() || WC_Payments_Features::is_upe_enabled() ) {
-			return;
-		}
-
 		$note_class = WC_Payment_Woo_Compat_Utils::get_note_class();
-		$note       = new $note_class();
-		$nonce      = wp_create_nonce( 'enable-upe' );
+		$note = new $note_class();
 
 		$note->set_title( __( 'Boost your sales by accepting new payment methods', 'woocommerce-payments' ) );
 		$note->set_content( __( 'Get early access to additional payment methods and an improved checkout experience, coming soon to WooCommerce Payments. <a href="https://docs.woocommerce.com/document/payments/" target="wcpay_upe_learn_more">Learn more</a>', 'woocommerce-payments' ) );
@@ -43,10 +89,11 @@ class WC_Payments_Notes_Additional_Payment_Methods {
 		$note->add_action(
 			self::NOTE_NAME,
 			__( 'Enable on your store', 'woocommerce-payments' ),
-			admin_url( 'admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments&action=enable-upe&_wpnonce=' . $nonce ),
+			admin_url( 'admin.php?page=wc-settings&action=' . self::ACTION_NAME ),
 			$note_class::E_WC_ADMIN_NOTE_UNACTIONED,
 			true
 		);
+		$note->add_nonce_to_action( self::NOTE_NAME, self::NONCE_NAME, '_wpnonce' );
 
 		return $note;
 	}

```

I'm not sure about the functionality in the `update_upe_enabled` static method.

The overall functionality of the note is great! 👍

harriswong commented on includes/notes/class-wc-payments-notes-additional-payment-methods.php:
```diff
@@ -0,0 +1,53 @@
+		$note->add_action(
+			self::NOTE_NAME,
+			__( 'Enable on your store', 'woocommerce-payments' ),
+			admin_url( 'admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments&action=enable-upe&_wpnonce=' . $nonce ),
```
Nice callout on `add_nonce_to_action`. I took a look at the [PR](https://github.com/woocommerce/woocommerce-admin/pull/6726/files) and it seems to do what I need to do here.

I tried using it but I got a function not found. After upgrading my woocommerce, it sparks a compatibility question. The `add_nonce_to_action` seems to be available [after WC 5.4](https://github.com/woocommerce/woocommerce/blob/6e208c9336db6b4804f4ba4782a6e64551aab2df/changelog.txt#L198), but WooCommerce Payments require [5.2+](https://github.com/Automattic/woocommerce-payments/blob/0c8683ae37148140d865eadcd8cd8914f4e2ad1e/readme.txt#L43) so there is a chance that the `add_nonce_to_action` won't be available to some. If this is the case, i'm hesitant to bump the version for this and would prefer to do `create_nonce()`, thoughts?

frosso replied:
That makes sense, thanks for taking a look! In that case, we can then just manually append the nonce through the string, like you did 👍

harriswong commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2227,4 +2229,30 @@ public function get_upe_available_payment_methods() {
+		WC_Payments_Features::update_upe_enabled( true );
+
+		$wcpay_settings_url = admin_url( 'admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments' );
+		wp_safe_redirect( $wcpay_settings_url );
```
Double checked https://developer.wordpress.org/reference/functions/wp_safe_redirect/, good call out!

One note, `exit;` will exit the unit test that I put in. Depending on how it will look like after some changes, the test may not be needed anymore and I will remove it.

frosso replied:
Gotcha, that makes sense - maybe `wp_die` could have a different behavior 🤷 not sure tho 😄

harriswong commented on includes/class-wc-payments-features.php:
```diff
@@ -89,4 +89,14 @@ public static function to_array() {
+	 * @param bool $enable True to enable UPE, false otherwise.
+	 */
+	public static function update_upe_enabled( $enable ) {
+		$enabled_flag = true ? 1 : 0 === $enable;
```
:/ This looks weird. This line flagged a yoda lint error and I might have yoda-ed it too much. Good catch!

harriswong commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -2227,4 +2229,30 @@ public function get_upe_available_payment_methods() {
+		}
+
+		// Enable UPE.
+		WC_Payments_Features::update_upe_enabled( true );
```
The idea of this function was to encapsulate the `_wcpay_feature_upe` flag name inside the feature class, following the approach of the getters inside that class. However, the latest `develop` introduces a `WC_Payments_Features::UPE_FLAG_NAME` const so this would suffice. Also agree on removing wrapper function that has no logic inside. 

> after enabling the feature flag, should we delete the note? 

I think it makes sense to.

harriswong commented on includes/notes/class-wc-payments-notes-additional-payment-methods.php:
```diff
@@ -0,0 +1,53 @@
+	 */
+	public static function get_note() {
+		// Show this notice only if UPE settings preview is disabled, and UPE flag is not enabled.
+		if ( ! WC_Payments_Features::is_upe_settings_preview_enabled() || WC_Payments_Features::is_upe_enabled() ) {
```
Thanks for the input. I ended up following some of the examples in src/note such as this ones: https://github.com/woocommerce/woocommerce-admin/blob/main/src/Notes/ChoosingTheme.php#L33

One thing to note is that `get_note()` or `[can_be_added()](https://github.com/woocommerce/woocommerce-admin/blob/2f4056f160c653bbb317492a258d6ac960511ce8/src/Notes/NoteTraits.php#L59-L61)` checks if the `$note` is falsy. If it's falsy, it does not proceed inside `possibly_add_note()`.

harriswong commented:
I updated the following changes:
- Moved `maybe_enable_upe_feature_flag()` back into `WC_Payments_Notes_Additional_Payment_Methods` class so that the note itself controls this method.
- During `add_woo_admin_notes` bootstrapping, run possibly_add_note() and maybe_enable_upe_feature_flag().
- Removed the function to update the `_wcpay_feature_upe` option since there is no logic inside anymore.
- Added a constant `NOTE_ACTION` for the `'enable-upe'` since it's used in a few places inside `WC_Payments_Notes_Additional_Payment_Methods`.
- Make use of `check_admin_referer` to check both nonce and a redirection from inside admin page.

frosso commented on includes/notes/class-wc-payments-notes-additional-payment-methods.php:
```diff
@@ -0,0 +1,81 @@
+			'wc-settings' !== $_GET['page'] ||
+			empty( $_GET['action'] ) ||
+			'enable-upe' !== $_GET['action'] ||
+			! check_admin_referer( self::NOTE_ACTION )
```
(very minor) we don't need to check for `if( ! check_admin_referer(...) )`, just doing `if (...) { return; } check_admin_referer( ... );` is sufficient

frosso replied:
UPDATE: maybe we do need to check - there is mixed usage. But this is good to go 👍

harriswong replied:
🚢

frosso reviewed (approved):
Thanks for considering the changes! Let's :shipit: !

harriswong commented:
Pushed another commit regarding some concerns about expired token for notes. Please take a look again!

frosso commented on includes/notes/class-wc-payments-notes-additional-payment-methods.php:
```diff
@@ -0,0 +1,83 @@
+		 * TODO: Add nonce to verify https://github.com/woocommerce/woocommerce-admin/pull/6726/
+		 * requires WC requirement is > 5.4.
+		 */
+		if (
+			empty( $_GET['page'] ) || // phpcs:disable WordPress.Security.NonceVerification.Recommended
+			'wc-settings' !== $_GET['page'] || // phpcs:disable WordPress.Security.NonceVerification.Recommended
+			empty( $_GET['action'] ) || // phpcs:disable WordPress.Security.NonceVerification.Recommended
+			'enable-upe' !== $_GET['action'] // phpcs:disable WordPress.Security.NonceVerification.Recommended
+		) {
+			return;
+		}
```
```suggestion
		if (
		    ! current_user_can( 'manage_woocommerce' ) ||
			empty( $_GET['page'] ) || // phpcs:disable WordPress.Security.NonceVerification.Recommended
			'wc-settings' !== $_GET['page'] || // phpcs:disable WordPress.Security.NonceVerification.Recommended
			empty( $_GET['action'] ) || // phpcs:disable WordPress.Security.NonceVerification.Recommended
			'enable-upe' !== $_GET['action'] // phpcs:disable WordPress.Security.NonceVerification.Recommended
		) {
			return;
		}
```

What are your thoughts on adding a capability check?

harriswong replied:
I was thinking that the admin notes was inside `admin_init` hooks,  `add_action( 'admin_init', [ __CLASS__, 'add_woo_admin_notes' ] );` . So then this method `maybe_enable_upe_feature_flag()` shouldn't be called if the user is not inside admin already. 

But being an admin doesn't imply they can `manage_woocommerce`. If anything, it maybe redundant but I see no harm in it. I can add this on top.

frosso reviewed (approved):
Still looking good 👍
